File: c/sort/quick_sort.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

#define N 10
int a[N] = {1, 6, 3, 8, 15, 6, 7, 3, 12, 10};

void quick_sort(int *a, int p, int q) {

    if (p >= q) {
        return;
    }

    int j = p;
    int povit = a[q];

    for (int i = p; i < q+1; i++) {
        if (a[i] < povit) {
            int tmp = a[i];
            a[i] = a[j];
            a[j] = tmp;

            j++;
        }
    }

    int tmp = a[j];
    a[j] = a[q];
    a[q] = tmp;

    quick_sort(a, p, j-1);
    quick_sort(a, j+1, q);
}

int main() {
    quick_sort(a, 0, N-1);

    for (int i = 0; i < N; i++)
        printf("%d, ", a[i]);
    printf("\n");
    return 0;
}

File: python/sort/quick_sort.py

a = [1, 4, 7, 3, 11, 9, 2, 6, 4, 7]
# a = [2, 3, 3, 7, 7, 9, 10, 12, 20]
# b = [1, 3, 4, 7, 10, 12, 12, 16, 17, 19]
# t = a + b

def quick_sort(a, p, q):

    if p == q:
        return

    j = p
    pivot = a[q]
    print('q = ', q)
    for i in range(p, q+1):
        print('a[%s] = %s' % (i, a[i]))
        if a[i] < pivot:
            tmp = a[i]
            a[i] = a[j]
            a[j] = tmp

            j += 1

    tmp = a[j]
    a[j] = a[q]
    a[q] = tmp

    print('p = %s, q = %s, j = %s' % (p, q, j))

    quick_sort(a, p, j-1)
    quick_sort(a, j+1, q)


def main():
    quick_sort(a, 0, len(a)-1)
    print(a)

if __name__ == '__main__':
    main()